### DISCUSSION TITLE
Symfony\Bundle\FrameworkBundle\Controller\AbstractController::setContainer(): issue

### DISCUSSION THREAD

**Discussion by ARY112** [symfony/symfony#46640] [Q&A] [answered]
I am getting below error after upadating to symfony 5.2 from 4.3
```
Symfony\Bundle\FrameworkBundle\Controller\AbstractController::setContainer(): Return value must be of type ?Psr\Container\ContainerInterface, 
App\SAMLBundle\Service\ContainerService returned.
```

Till Version 4.3 
```php
# symfony/src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php
public function setContainer(ContainerInterface $container)
 ```
After Version 4.4 
```php
public function setContainer(ContainerInterface $container): ?ContainerInterface
```

SampleCode from Bundle DefaultController
Previous Code which was working in version 4.3

```php
    public function __construct(ContainerInterface $container)
    {
        $serviceObj = $container->get('saml_container_service');
        $this->container = ContainerSingleton::setContainer($serviceObj);
    }
```	
Had to modify as per symfony version 5.2
```php
    public function __construct(ContainerInterface $container): ?ContainerInterface
    {
        $serviceObj = $container->get('saml_container_service');
        $this->container = ContainerSingleton::setContainer($serviceObj);
    }
```
Can some one suggest what changes i should make it to work.

**Comment by lyrixx** [comment, marked as answer]
(Hi, I edited you message to add syntax highlighting) 

In Symfony 5.2 you may return a the previous container, but this is not mandatory
https://github.com/symfony/symfony/blob/5.2/src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php#L66

But be careful, in your last snippet, you are edditing the `__construct()` method and not the `setContainer()` method. And a controller does not return anything.

---

So it looks like you are injecting in Container in your constructor, and you replace the container in the parent (AbstractController) by a `ContainerSingleton`. 

Then symfony, during the loading of your controller will inject a very special container. It's a service locator, that hold only few references to very [specific and common services](https://github.com/symfony/symfony/blob/5.2/src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php#L88-L106) .

And during this loading, the previous container will be return and this throw an error because the previous container was not an instance of `ContainerInterface` but an instance of `ContainerSingleton`

You have many way to fix it. I let you with some possibility, take the one you prefer 

* avoid using a singleton in Symfony. This is considerer as a bad practice. Inject all services you need in your Controller, and do not use the `container` property
* use another property name, like `containerSingleton` instead of `container` (**this is the easiest one)
* keep using `container` but declare a **private** property in your controller (not tested !)
